README.txt
googlesitemap.common
====================

This core package includes sitemap generation common utilities with `sitemap indexes support`. 
This way you will be more Google-friendly and you will be able to use sitemaps with very large sites.

googlesitemap.common is not intended to be a standalone package. If you want to write a specific sitemap support you may start from here using its utilities.

Authors
-------

- Davide Moro <davide.moro@redomino.com>




docs/HISTORY.txt
Changelog
=========

1.4 (unreleased)
----------------

- Nothing changed yet.


1.3 (2012-05-04)
----------------

- removed paster plugins [sithmel]


1.2 (2012-04-02)
----------------

- added MANIFEST.in [sithmel]


1.1 (2012-04-02)
----------------

- Nothing changed yet.


1.0 (2011-04-15)
----------------

* Initial release



googlesitemap/common/tests.py
import unittest

from zope.testing import doctestunit
from zope.component import testing
from Testing import ZopeTestCase as ztc

from Products.Five import zcml
from Products.Five import fiveconfigure
from Products.PloneTestCase import PloneTestCase as ptc
from Products.PloneTestCase.layer import PloneSite
ptc.setupPloneSite()

import googlesitemap.common

class TestCase(ptc.PloneTestCase):
    class layer(PloneSite):
        @classmethod
        def setUp(cls):
            fiveconfigure.debug_mode = True
            zcml.load_config('configure.zcml',
                             googlesitemap.common)
            fiveconfigure.debug_mode = False

        @classmethod
        def tearDown(cls):
            pass


def test_suite():
    return unittest.TestSuite([

        # Unit tests
        doctestunit.DocFileSuite(
            'sitemap.txt', package='googlesitemap.common',
            setUp=testing.setUp, tearDown=testing.tearDown),

        #doctestunit.DocTestSuite(
        #    module='googlesitemap.common.mymodule',
        #    setUp=testing.setUp, tearDown=testing.tearDown),


        # Integration tests that use PloneTestCase
        #ztc.ZopeDocFileSuite(
        #    'README.txt', package='googlesitemap.common',
        #    test_class=TestCase),

        #ztc.FunctionalDocFileSuite(
        #    'browser.txt', package='googlesitemap.common',
        #    test_class=TestCase),

        ])

if __name__ == '__main__':
    unittest.main(defaultTest='test_suite')


googlesitemap/common/tests/base.py
# Authors: Davide Moro <davide.moro@redomino.com> and contributors (see docs/CONTRIBUTORS.txt)
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License version 2 as published
# by the Free Software Foundation.
# 
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
# 
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA
# 02111-1307, USA.
"""Test setup for integration and functional tests.

When we import PloneTestCase and then call setupPloneSite(), all of
Plone's products are loaded, and a Plone site will be created. This
happens at module level, which makes it faster to run each test, but
slows down test runner startup.
"""

from Products.Five import zcml
from Products.Five import fiveconfigure

from Testing import ZopeTestCase as ztc

from Products.PloneTestCase import PloneTestCase as ptc
from Products.PloneTestCase.layer import onsetup

# When ZopeTestCase configures Zope, it will *not* auto-load products
# in Products/. Instead, we have to use a statement such as:
#   ztc.installProduct('SimpleAttachment')
# This does *not* apply to products in eggs and Python packages (i.e.
# not in the Products.*) namespace. For that, see below.
# All of Plone's products are already set up by PloneTestCase.


@onsetup
def setup_product():
    """Set up the package and its dependencies.

    The @onsetup decorator causes the execution of this body to be
    deferred until the setup of the Plone site testing layer. We could
    have created our own layer, but this is the easiest way for Plone
    integration tests.
    """
    # Load the ZCML configuration for the example.tests package.
    # This can of course use <include /> to include other packages.

    fiveconfigure.debug_mode = True

    import plone.browserlayer
    zcml.load_config('configure.zcml', plone.browserlayer)
    import googlesitemap.common
    zcml.load_config('configure.zcml', googlesitemap.common)
    import googlesitemap.common.tests
    zcml.load_config('testing.zcml', googlesitemap.common.tests)
    fiveconfigure.debug_mode = False

    # fix important, we avoid an error on setupPloneSite


    # We need to tell the testing framework that these products
    # should be available. This can't happen until after we have loaded
    # the ZCML. Thus, we do it here. Note the use of installPackage()
    # instead of installProduct().
    # This is *only* necessary for packages outside the Products.*
    # namespace which are also declared as Zope 2 products, using
    # <five:registerPackage /> in ZCML.

    # We may also need to load dependencies, e.g.:
    #   ztc.installPackage('borg.localrole')

    ztc.installPackage('plone.browserlayer')
    ztc.installPackage('googlesitemap.common')

# The order here is important: We first call the (deferred) function
# which installs the products we need for this product. Then, we let
# PloneTestCase set up this product on installation.

setup_product()
ptc.setupPloneSite(products=['googlesitemap.common',],) 


class TestCase(ptc.PloneTestCase):
    """We use this base class for all the tests in this package. If
    necessary, we can put common utility or setup code in here. This
    applies to unit test cases.
    """


class FunctionalTestCase(ptc.FunctionalTestCase):
    """We use this class for functional integration tests that use
    doctest syntax. Again, we can put basic common utility or setup
    code in here.
    """

    def afterSetUp(self):
        roles = ('Member', 'Contributor')
        self.portal.portal_membership.addMember('contributor',
                                                'secret',
                                                roles, [])


googlesitemap/common/tests/test_sitemap.py
# Authors: Davide Moro <davide.moro@redomino.com> and contributors (see docs/CONTRIBUTORS.txt)
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License version 2 as published
# by the Free Software Foundation.
# 
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
# 
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA
# 02111-1307, USA.

from gzip import GzipFile
from StringIO import StringIO

from zope.component import getMultiAdapter
from zope.publisher.interfaces import INotFound

from Products.CMFCore.utils import getToolByName

from googlesitemap.common.tests.base import TestCase


class SiteMapTestCase(TestCase):
    """base test case with convenience methods for all sitemap tests"""

    def afterSetUp(self):
        super(SiteMapTestCase, self).afterSetUp()
        self.wftool = getToolByName(self.portal, 'portal_workflow')

        # we need to explizitly set a workflow cause we can't rely on the
        # test environment.
        # `instance test -m plone.app.layout`:
        # wftool._default_chain == 'simple_publication_workflow'
        # `instance test -m plone.app`:
        # wftool._default_chain == 'plone_workflow'
        self.wftool.setChainForPortalTypes(['Document'],
                                           'simple_publication_workflow')

        self.site_properties = getToolByName(
            self.portal, 'portal_properties').site_properties
        self.site_properties.manage_changeProperties(enable_sitemap=True)

        #setup private content that isn't accessible for anonymous
        self.loginAsPortalOwner()
        self.portal.invokeFactory(id='private', type_name='Document')
        private = self.portal.private
        self.assertTrue('private' == self.wftool.getInfoFor(private,
                                                            'review_state'))
        
        #setup published content that is accessible for anonymous
        self.portal.invokeFactory(id='published', type_name='Document')
        published = self.portal.published
        self.wftool.doActionFor(published, 'publish')
        self.assertTrue('published' == self.wftool.getInfoFor(published,
                                                              'review_state'))

        #setup pending content that is accessible for anonymous
        self.portal.invokeFactory(id='pending', type_name='Document')
        pending = self.portal.pending
        self.wftool.doActionFor(pending, 'submit')
        self.assertTrue('pending' == self.wftool.getInfoFor(pending,
                                                            'review_state'))

        self.portal.invokeFactory(id='file1', type_name='File')
        self.file1 = self.portal.file1
        self.portal.invokeFactory(id='image1', type_name='Image')
        self.image1 = self.portal.image1

        self.logout()

    @property
    def sitemap(self):
        return getMultiAdapter((self.portal, self.portal.REQUEST),
                               name='sitemapindex.xml.gz')
        
    def uncompress(self, sitemapdata):
        sio = StringIO(sitemapdata)
        unziped = GzipFile(fileobj=sio)
        xml = unziped.read()
        unziped.close()
        return xml

    def test_layers(self):
        """ Browser layers setup """
        from googlesitemap.common.interfaces import ISitemapLayer

        from plone.browserlayer.utils import registered_layers
        # our layer is correctly applied
        self.assertTrue(ISitemapLayer in registered_layers())

    def test_view(self):
        """ If the sitemap class is right """
        from googlesitemap.common.sitemap import SiteMapCommonView
        # this test fails, we need to fix the tests setup
        # I've created a fake testing.zcml to fix this
        self.assertTrue(isinstance(self.sitemap, SiteMapCommonView))

    def test_disabled(self):
        '''
        If the sitemap is disabled throws a 404 error.
        '''
        self.site_properties.manage_changeProperties(enable_sitemap=False)
        try:
            self.sitemap()
        except Exception, e:
            # zope2 and repoze.zope2 use different publishers and raise
            # different exceptions. but both implement INotFound.
            self.assertTrue(INotFound.providedBy(e))
        else:
            self.fail('The disabled sitemap view has to raise NotFound!')
            
    def test_authenticated_before_anonymous(self):
        '''
        Requests for the sitemap by authenticated users are not cached.
        anomymous users get a uncached sitemap that only contains content
        that they are supposed to see.
        '''

        # first round as an authenticated (manager)
        self.loginAsPortalOwner()
        xml = self.uncompress(self.sitemap())
        self.assertTrue('<loc>http://nohost/plone/private</loc>' in xml)
        self.assertTrue('<loc>http://nohost/plone/pending</loc>' in xml)
        self.assertTrue('<loc>http://nohost/plone/published</loc>' in xml)

        # second round as anonymous
        self.logout()
        xml = self.uncompress(self.sitemap())
        self.assertFalse('<loc>http://nohost/plone/private</loc>' in xml)
        self.assertFalse('<loc>http://nohost/plone/pending</loc>' in xml)
        self.assertTrue('<loc>http://nohost/plone/published</loc>' in xml)

    def test_anonymous_before_authenticated(self):
        '''
        Requests for the sitemap by anonymous users are cached.
        authenticated users get a uncached sitemap. Test that the cached
        Sitemap is not delivered to authenticated users.
        '''

        # first round as anonymous
        xml = self.uncompress(self.sitemap())
        self.assertFalse('<loc>http://nohost/plone/private</loc>' in xml)
        self.assertFalse('<loc>http://nohost/plone/pending</loc>' in xml)
        self.assertTrue('<loc>http://nohost/plone/published</loc>' in xml)

        # second round as an authenticated (manager)
        self.loginAsPortalOwner()
        xml = self.uncompress(self.sitemap())
        self.assertTrue('<loc>http://nohost/plone/private</loc>' in xml)
        self.assertTrue('<loc>http://nohost/plone/pending</loc>' in xml)
        self.assertTrue('<loc>http://nohost/plone/published</loc>' in xml)

    def test_changed_catalog(self):
        '''
        The sitemap is generated from the catalog. If the catalog changes, a new
        sitemap has to be generated.
        '''

        xml = self.uncompress(self.sitemap())
        self.assertFalse('<loc>http://nohost/plone/pending</loc>' in xml)

        # changing the workflow state 
        self.loginAsPortalOwner()
        pending = self.portal.pending
        self.wftool.doActionFor(pending, 'publish')
        self.logout()

        xml = self.uncompress(self.sitemap())
        self.assertTrue('<loc>http://nohost/plone/pending</loc>' in xml)

        #removing content
        self.loginAsPortalOwner()
        self.portal.manage_delObjects(['published',])
        self.logout()

        xml = self.uncompress(self.sitemap())
        self.assertFalse('<loc>http://nohost/plone/published</loc>' in xml)        

    def test_image_file(self):
        """ Image and files are handled correctly? """
        xml = self.uncompress(self.sitemap())
        self.assertTrue(self.image1.absolute_url() in xml)
        self.assertTrue(self.file1.absolute_url() in xml)

        # both file and file/view in sitemap
        file_url = self.file1.absolute_url()
        self.assertTrue('<loc>%s</loc>' % file_url in xml)
        self.assertTrue('<loc>%s/view</loc>' % file_url in xml)

        # plain images should not be included in sitemaps, use sitemap for images!
        image_url = self.image1.absolute_url()
        self.assertFalse('<loc>%s</loc>' % image_url in xml)
        self.assertTrue('<loc>%s/view</loc>' % image_url in xml)

    def test_lastmod(self):
        """ 
            Seems to be better having spaces in lastmod.
            It is undocumented but it seems better this format
              <lastmod> DATE </lastmod>  
            instead of:
              <lastmod>DATE</lastmod>  

            DATE should be in ZULU time format ('2010-12-14T10:53:21Z')
        """
        xml = self.uncompress(self.sitemap())
        zulu_time = self.image1.modified().HTML4()
        self.assertTrue('<lastmod> %s </lastmod>' % zulu_time in xml)
        self.assertFalse('<lastmod>%s</lastmod>' % zulu_time in xml)


def test_suite():
    from unittest import defaultTestLoader
    return defaultTestLoader.loadTestsFromName(__name__)


googlesitemap/__init__.py
# Authors: Davide Moro <davide.moro@redomino.com> and contributors (see docs/CONTRIBUTORS.txt)
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License version 2 as published
# by the Free Software Foundation.
# 
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
# 
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA
# 02111-1307, USA.
# See http://peak.telecommunity.com/DevCenter/setuptools#namespace-packages
try:
    __import__('pkg_resources').declare_namespace(__name__)
except ImportError:
    from pkgutil import extend_path
    __path__ = extend_path(__path__, __name__)


googlesitemap/common/interfaces.py
# Authors: Davide Moro <davide.moro@redomino.com> and contributors (see docs/CONTRIBUTORS.txt)
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License version 2 as published
# by the Free Software Foundation.
# 
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
# 
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA
# 02111-1307, USA.

from zope.interface import Interface
from zope.interface import Attribute

class ISitemapLayer(Interface):
    """ Layer interface """

class ISiteMapView(Interface):
    """ Sitemap view """

    template = Attribute("""Template used for sitemap generation""")
    indextemplate = Attribute("""Template used for sitemap indexes""")

    maxlen = Attribute("""The maximum number of items for sitemap""")
    query_dict = Attribute("""The catalog query used for get sitemap elements""")
    filename = Attribute("""The generated sitemap's filename""")
    enable_sitemap = Attribute("""Sitemap generation available only if enable_sitemap is enabled""")

    def sitemaps():
        """Get sitemaps data when using indexes of sitemaps"""

    def getStartEnd():
        """ Get window slice of brains used for indexes of sitemaps"""

    def objects():
        """Returns the data to create the sitemap. It may be a portion of all existing objects when we need to
           generate sitemap indexes.
        """

    def generate():
        """ Generates the cached Gzipped sitemap.
            You can just override this method for change caching policy (different kind of sitemaps may have
            different settings).
        """

    def __call__():
        """Checks if the sitemap feature is enabled and returns it.
           It may returns a standard sitemap or an index of sitemaps when where are more than maxlen objects
           on our site.
        """






googlesitemap/common/sitemap.py
# Authors: Davide Moro <davide.moro@redomino.com> and contributors (see docs/CONTRIBUTORS.txt)
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License version 2 as published
# by the Free Software Foundation.
# 
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
# 
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA
# 02111-1307, USA.

from gzip import GzipFile
from cStringIO import StringIO

from DateTime import DateTime
from zope.interface import implements
from zope.component import getMultiAdapter
from zope.publisher.interfaces import NotFound

from plone.memoize import ram
from plone.memoize.instance import memoize

from Products.Five import BrowserView
from Products.Five.browser.pagetemplatefile import ViewPageTemplateFile
from Products.CMFCore.utils import getToolByName

from googlesitemap.common.interfaces import ISiteMapView
from googlesitemap.common import config


def _render_defaultcachekey(fun, self):
    # Cache by filename
    mtool = getToolByName(self.context, 'portal_membership')
    if not mtool.isAnonymousUser():
        raise ram.DontCache

    url_tool = getToolByName(self.context, 'portal_url')
    catalog = getToolByName(self.context, 'portal_catalog')
    counter = catalog.getCounter()
    return '%s/%s/%s/%s' % (url_tool(), self.filename, counter, str(self.index))


class SiteMapCommonView(BrowserView):
    """ Base class for build Sitemaps """
    implements(ISiteMapView)

    template = ViewPageTemplateFile('sitemap.xml')
    indextemplate = ViewPageTemplateFile('sitemapindex.xml')

    def __init__(self, context, request):
        self.context = context
        self.request = request
        self.index = getattr(self.request, 'index', None)

    @property
    def maxlen(self):
        return config.MAXLEN

    @property
    def query_dict(self):
        return {'Language': 'all', 
                'sort_on': 'Date',
                'sort_order':'reverse',
               }

    @property
    def filename(self):
        return 'sitemap.xml'

    @property
    def enable_sitemap(self):
        sp = getToolByName(self.context, 'portal_properties').site_properties
        return sp.enable_sitemap

    @memoize
    def portal_url(self):
        return getMultiAdapter((self.context, self.request), name=u"plone_portal_state").portal_url()

    def sitemaps(self):
        """ Get sitemaps data when using indexes of sitemaps """
        for current in enumerate(self._catalogbrains()[0::self.maxlen]):
            index = current[0]
            item = current[1]
           
            maxdate = DateTime(item.Date).ISO8601()
            url = '%s/%s?index=%d' % (self.portal_url(), self.filename, index)
            yield {'maxdate':maxdate, 'url':url}

# TODO: re-enable memoize without breaking tests
#    @memoize
    def _catalogbrains(self):
        """Returns the data to create the sitemap. Max items = 1000 * maxlen using sitemap indexes.
           maxlen depends on the specific sitemap (standard sitemap, video, news, etc).
        """
        catalog = getToolByName(self.context, 'portal_catalog')
        return catalog.searchResults(**self.query_dict)[:self.maxlen*1000]

    def getStartEnd(self):
        """ Get window slice of brains """
        start = None
        end = None
        if self.index is not None:
            start = int(self.index)*self.maxlen
            end = start + self.maxlen
        return (start, end)

    def _slicecatalogbrains(self):
        """ Get the right window of brains """
        start, end = self.getStartEnd()
        if start is not None and end is not None:
            catalog_brains = self._catalogbrains()[start:end]
        else:
            catalog_brains = self._catalogbrains()
        return catalog_brains


    def objects(self):
        """Returns the data to create the sitemap."""
        catalog_brains = self._slicecatalogbrains()

        for item in catalog_brains:
            if item.portal_type in ['Image']:
                yield {
                    'loc': '%s/view' % item.getURL(),
                    'lastmod': item.modified.HTML4(),
                    #'changefreq': 'always', # hourly/daily/weekly/monthly/yearly/never
                    #'prioriy': 0.5, # 0.0 to 1.0
                }
            else:
                if item.portal_type in ['File']:
                    yield {
                        'loc': '%s/view' % item.getURL(),
                        'lastmod': item.modified.HTML4(),
                        #'changefreq': 'always', # hourly/daily/weekly/monthly/yearly/never
                        #'prioriy': 0.5, # 0.0 to 1.0
                    }
                yield {
                    'loc': item.getURL(),
                    'lastmod': item.modified.HTML4(),
                    #'changefreq': 'always', # hourly/daily/weekly/monthly/yearly/never
                    #'prioriy': 0.5, # 0.0 to 1.0
                }

    def _uncachedgenerate(self):
        """ Generates the Gzipped sitemap uncached data """
        len_brains = len(self._catalogbrains())

        if self.index is None:
            # no index specified in the url
            if len_brains < self.maxlen:
                # ok, we have few items, let's generate the standard sitemap
                xml = self.template()
            else:
                # a lot of items, let's generate a sitemap index
                xml = self.indextemplate()
        elif int(self.index)*self.maxlen >= len_brains:
            # bad index specified
            raise NotFound(self.context, '%s-%s' % (self.index, self.filename), self.request)
        else:
            # index specified in the url
            xml = self.template()

        if self.index is not None:
            filename = "%s-%s" % (self.index, self.filename)
        else:
            filename = self.filename

        fp = StringIO()
        gzip = GzipFile(filename, 'w', 9, fp)
        gzip.write(xml)
        gzip.close()
        data = fp.getvalue()
        fp.close()
        return data

    @ram.cache(_render_defaultcachekey)
    def generate(self):
        """ Generates the cached Gzipped sitemap.
            Override just this method for change caching policy
        """
        return self._uncachedgenerate()

    def __call__(self):
        """Checks if the sitemap feature is enabled and returns it."""
        if not self.enable_sitemap:
            # sitemap enabled
            raise NotFound(self.context, '%s' % self.filename, self.request)

        self.request.response.setHeader('Content-Type',
                                        'application/octet-stream')
        return self.generate()


